source/common/json/json_loader.cc

```diff
@@ -2,242 +2,669 @@
 
 #include <cstdint>
 #include <fstream>
+#include <limits>
+#include <sstream>
+#include <stack>
 #include <string>
+#include <unordered_map>
 #include <vector>
 
+#include "common/common/assert.h"
+#include "common/common/utility.h"
+#include "common/filesystem/filesystem_impl.h"
+
 #include "spdlog/spdlog.h"
 
 // Do not let RapidJson leak outside of this file.
 #include "rapidjson/document.h"
 #include "rapidjson/error/en.h"
-#include "rapidjson/istreamwrapper.h"
+#include "rapidjson/reader.h"
 #include "rapidjson/schema.h"
+#include "rapidjson/stream.h"
 #include "rapidjson/stringbuffer.h"
 #include "rapidjson/writer.h"
 
 namespace Envoy {
 namespace Json {
 
-/**
- * Implementation of Object.
+namespace {
+/*
+ * Internal representation of Object.
  */
-class ObjectImplBase : public Object {
+class Field;
+typedef std::shared_ptr<Field> FieldSharedPtr;
+
+class Field : public Object, public std::enable_shared_from_this<Field> {
 public:
-  ObjectImplBase(const rapidjson::Value& value, const std::string& name)
-      : name_(name), value_(value) {}
+  void setLineNumberStart(uint64_t line_number) { line_number_start_ = line_number; }
+  void setLineNumberEnd(uint64_t line_number) { line_number_end_ = line_number; }
 
-  std::vector<ObjectPtr> asObjectArray() const override {
-    if (!value_.IsArray()) {
-      throw Exception(fmt::format("'{}' is not an array", name_));
-    }
+  // Container factories for handler.
+  static FieldSharedPtr createObject() { return FieldSharedPtr{new Field(Type::Object)}; }
```

Comment:
let's use std::make_shared here and everywhere else we are making new shared_ptr inline.